yapapi/log.py

```diff
@@ -0,0 +1,271 @@
+    and output some summary information.
+
+    The optional `wrapped_emitter` argument can be used for chaining event
+    emitters: each event logged with `log()` is first passed to
+    `wrapped_emitter`.
+
+    For example, with the following setup, each event emitted by `engine`
+    will be logged by `log_event_json`, and additionally, certain events
+    will cause summary messages to be logged.
+    ```
+        detailed_logger = log_event_json
+        summary_logger = SummaryLogger(wrapped_emitter=detailed_logger).log
+        engine = Engine(..., event_emitter=summary_logger)
+    ```
+    """
+
+    logger = logging.getLogger("yapapi.summary")
+
+    # Generates subsequent numbers, for use in generated provider names
+    numbers: Iterator[int]
+
+    # Start time of the computation
+    start_time: float
+
+    # Set of confirmed proposal ids
+    confirmed_proposals: Set[str]
+
+    # Maps agreement ids to provider names
+    agreement_provider_name: Dict[str, str]
+
+    # Set of agreements confirmed by providers
+    confirmed_agreements: Set[str]
+
+    # Maps task id to task data
+    task_data: Dict[str, Any]
+
+    # Maps a provide name to the list of task ids computed by the provider
+    provider_tasks: Dict[str, List[str]]
+
+    # Map a provider name to the sum of amounts in this provider's invoices
+    provider_cost: Dict[str, float]
+
+    # Has computation finished?
+    finished: bool
+
+    def __init__(self, wrapped_emitter: Optional[Callable[[EventType], None]] = None):
+        """Create a SummaryLogger."""
+
+        self._wrapped_emitter = wrapped_emitter
+        self.numbers: Iterator[int] = itertools.count(1)
+        self._reset()
+
+    def _reset(self) -> None:
+        """Reset all information aggregated by this logger."""
+
+        self.start_time = time.time()
+        self.confirmed_proposals = set()
+        self.agreement_provider_name = {}
+        self.confirmed_agreements = set()
+        self.task_data = {}
+        self.provider_tasks = defaultdict(list)
+        self.provider_cost = {}
+        self.finished = False
+        self.error_occurred = False
+
+    def _print_total_cost(self) -> None:
+
+        if not self.finished:
+            return
+
+        provider_names = set(self.provider_tasks.keys())
+        if set(self.provider_cost).issuperset(provider_names):
+            total_cost = sum(self.provider_cost.values(), 0.0)
+            self.logger.info("Total cost: %s", total_cost)
+
+    def log(self, event: EventType) -> None:
+        """Register an event."""
+
+        if self._wrapped_emitter:
+            self._wrapped_emitter(event)
+
+        if self.error_occurred:
+            return
+
+        try:
+            self._handle(event)
+        except Exception:
+            self.logger.exception("SummaryLogger entered invalid state")
+            self.error_occurred = True
+
+    def _handle(self, event: EventType):
+
+        if isinstance(event, Event.ComputationStarted):
```

Comment:
maybe this if-ladder shows that this model is not very handy.

Reply:
Perhaps it is not, but I don't see any good alternative for this sequence of `if isinstance`s. We could use the visitor pattern but it has its overhead and so far we only need it once.